bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgerMetadataIndex.java

```diff
@@ -228,20 +228,28 @@ public void flush() throws IOException {
 
     public void removeDeletedLedgers() throws IOException {
         LongWrapper key = LongWrapper.get();
+        final byte[] startKey = new byte[8];
+        final byte[] endKey = new byte[8];
 
         int deletedLedgers = 0;
         while (!pendingDeletedLedgers.isEmpty()) {
             long ledgerId = pendingDeletedLedgers.poll();
             key.set(ledgerId);
             ledgersDb.delete(key.array);
-            deletedLedgers++;
+            if (deletedLedgers++ == 0) {
+                System.arraycopy(key.array, 0, startKey, 0, startKey.length);
+            }
+            System.arraycopy(key.array, 0, endKey, 0, endKey.length);
         }
 
         if (log.isDebugEnabled()) {
             log.debug("Persisting deletes of ledgers {}", deletedLedgers);
         }
 
         ledgersDb.sync();
+        if (deletedLedgers != 0) {
+            ledgersDb.compact(startKey, endKey);
```

Comment:
as I understood, this method runs on each checkpoint.
How expensive is this compaction in terms of time?
IIRC checkpointing happens on entry log rotation/gets into the addEntry latency, how much this increases p99.9 under load?

I am not familiar with RocksDB nuances. Looking at https://github.com/facebook/rocksdb/wiki/Manual-Compaction

> In case of universal and FIFO compaction styles, the begin and end arguments are ignored and all files are compacted. Also, files in each level are compacted and left in the same level. For leveled compaction style, all files containing keys in the given range are compacted to the last level containing files. 

Sounds like this is potentially expensive operation creating additional IO.

Would it be better run on a separate thread at lower frequency, better yet as a part of bookie's garbage collection?

Reply:
The two methods I modified are only called by the sync thread, so already in the background.

As far as additional IO: potentially yes.  Each compaction will need to flush the memtable and then read each SST that contains a key in the range. Each key is likely present in at most two SSTs: the delete tombstone and the original put. All of the affected SSTs are then rewritten.  This is background work with regards to the get/put/delete operations but will stall the sync thread.  In the case where writes to ledgers are coming in fast, most of the keys should be adjacent or nearly so.